packages/docusaurus-mdx-loader/src/remark/toc/__tests__/index.test.ts

```diff
@@ -30,6 +30,14 @@ describe('toc remark plugin', () => {
     expect(result).toMatchSnapshot();
   });
 
+  // A very implicit API: we allow users to hand-write the toc variable. It will
+  // get overwritten in most cases, but until we find a better way, better keep
+  // supporting this
+  it('does not overwrite TOC var if no TOC', async () => {
+    const result = await processFixture('no-heading-with-toc-export');
+    expect(result).toMatchSnapshot();
+  });
+
```

Comment:
@Josh-Cena I don't understand why this behavior was implemented.

As far as I understand:

Manually exported toc will be overriden by remark toc if there are headings:

```mdx
## heading

export const toc = [{}]
```

Manually exported toc won't be overriden by remark toc if there's no heading

```mdx

export const toc = [{}]
```

Why would we want 2 different behaviors here? 

I'd like to get rid of one and always let the user override the toc with an explicit export.

Asking for https://github.com/facebook/docusaurus/pull/9684

Reply:
@slorber: It was implemented to support hand-writing TOC in very edge cases, such as importing partials, using exclusively JSX headings, and such. I think I implemented it this way, because if a user both has hand-written and existing TOC, it is more likely that they should be merged, but I'm okay with always-overriding behavior too. I just thought this is the least surprising for the time being.

Reply:
Thanks @Josh-Cena 


After my refactor the logic is much easier to understand, and now feels a bit awkward to me.

<img width="709" alt="CleanShot 2024-01-19 at 01 15 57@2x" src="https://github.com/facebook/docusaurus/assets/749374/4219ab3b-fc37-48bf-afe8-84f9fc98695a">

I'll probably update this behavior and a few other things in another PR, for now I kept the existing behavior although 2 things surprised me:
- exporting the toc after the last import (instead of always at the end)
- overriding user explicit toc only if any heading is found

Reply:
> exporting the toc after the last import (instead of always at the end)

This has always been the case, which surprised me too because it's extra work.

> overriding user explicit toc only if any heading is found

Which to me is a slight improvement over always overriding explicit TOC, although I agree always respecting explicit TOC is probably the most reasonable.

Reply:
I think it has always remained an implementation detail, and while we did recommend using `export toc = [...toc1, ...toc2]` for the "barrel of partials" case before, this use case will keep working. I don't think anyone will expect their explicit TOC to be overridden.